src/routes/notifications/v1/notifications.controller.ts

```diff
@@ -11,27 +12,119 @@ import { RegisterDeviceDto } from '@/routes/notifications/v1/entities/register-d
+  ) {}
 
   @ApiOkResponse()
   @Post('register/notifications')
   @HttpCode(200)
   async registerDevice(
     @Body() registerDeviceDto: RegisterDeviceDto,
   ): Promise<void> {
-    return this.notificationsService.registerDevice(registerDeviceDto);
+    await this.notificationsService.registerDevice(registerDeviceDto);
+
+    // Compatibility with V2
+    // @TODO Remove NotificationModuleV2 after all clients have migrated and compatibility is no longer needed.
+    const compatibleV2Requests =
+      await this.createV2RegisterDto(registerDeviceDto);
+
+    for (const compatibleV2Request of compatibleV2Requests) {
+      await this.notificationServiceV2.upsertSubscriptions(compatibleV2Request);
+    }
+  }
+
+  private async createV2RegisterDto(args: RegisterDeviceDto): Promise<
+    Array<{
+      upsertSubscriptionsDto: UpsertSubscriptionsDto;
+      authPayload: AuthPayload;
+    }>
+  > {
+    const safeV2Array: Array<{
+      authPayload: AuthPayload;
+      upsertSubscriptionsDto: UpsertSubscriptionsDto & {
+        signature: `0x${string}`;
+      };
+    }> = [];
+
+    const safesV1Registrations = args.safeRegistrations;
+
+    for (const safeV1Registration of safesV1Registrations) {
+      if (safeV1Registration.safes.length) {
+        const safeV2: {
+          authPayload: AuthPayload;
+          upsertSubscriptionsDto: UpsertSubscriptionsDto & {
+            safes: Array<UpsertSubscriptionsSafesDto>;
+            signature: `0x${string}`;
+          };
+        } = {
+          upsertSubscriptionsDto: {
+            cloudMessagingToken: args.cloudMessagingToken,
+            deviceType: args.deviceType,
+            deviceUuid: (args.uuid as UUID) || undefined,
+            safes: [],
+            signature: safeV1Registration.signatures[0] as `0x${string}`,
+          },
+          authPayload: new AuthPayload(),
+        };
+        for (const safeAddresses of safeV1Registration.safes) {
+          safeV2.upsertSubscriptionsDto.safes.push({
+            address: safeAddresses as `0x${string}`,
+            chainId: safeV1Registration.chainId,
+            notificationTypes: Object.values(NotificationType),
+          });
+        }
+        safeV2Array.push(safeV2);
+      }
+    }
+
+    for (const [index, safeV2] of safeV2Array.entries()) {
+      const safeAddresses = safeV2.upsertSubscriptionsDto.safes.map(
+        (safeV2Safes) => safeV2Safes.address,
+      );
+
+      const recoveredAddress = await recoverMessageAddress({
+        message: `gnosis-safe${args.timestamp}${args.uuid}${args.cloudMessagingToken}${safeAddresses.sort().join('')}`,
+        signature: safeV2.upsertSubscriptionsDto.signature,
+      });
+
+      safeV2.authPayload.chain_id =
+        safeV2.upsertSubscriptionsDto.safes[0].chainId;
+      safeV2.authPayload.signer_address = recoveredAddress;
+
+      safeV2Array[index].authPayload = safeV2.authPayload;
+    }
+
+    return safeV2Array;
```

Comment:
This shares a lot of similarity with `create-registration-v2.dto.builder.ts`. Maybe we could extract some helpers. My comments regarding the former stand here too.

Reply:
The `create-registration-v2.dto.builder.ts` file is located in the `__tests__` folder.
I’d recommend not modifying it, as the duplication is intentional and serves a specific purpose for testing.

Reply:
I agree that using a test-based function here doesn't make sense. We could create a helper that is used both here and there though.

Reply:
I prefer to keep the tests somewhat isolated. In this case, duplicating the logic within the tests ensures that any changes here will cause the tests to fail, making us immediately aware of any modifications and possible errors. What do you think?

Reply:
If you think it best, I'm happy to keep it as is. What do you think about implementing my [previous suggestion regarding the `AuthPayload`](https://github.com/safe-global/safe-client-gateway/pull/2063#discussion_r1828284192) here though?

Reply:
Please look at my previous comment.